### ISSUE TITLE
why can't emit in a sub-thread in a Bolt

### ISSUE THREAD

**Issue by compassur** [Parsely/streamparse#239]
i'm using a Bolt for maintaining a hash-table and want to auto-emit it periodicity.
but if i use a Timer thread in a Bolt to periodicity emit the hash-table, it raises exception.

java.lang.Exception: Shell Process Exception: Python JSONDecodeError raised
java.lang.RuntimeException: backtype.storm.multilang.NoOutputException: Pipe to subprocess seems to be broken! No output read.

thanks

**Comment by dan-blanchard** [comment]
Emitting from a thread shouldn't be a problem, as we even provide a class, `TicklessBatchingBolt`, which uses a timer thread and processes/emits periodically.  Without more of your log output, it's hard to say what exactly went wrong.

**Comment by compassur** [comment]
in in my bolt i use Timer, it is sometion like that

```python
from threading import Timer
class MyBolt(bolt):
  def initialize(self, storm_conf, context):
      Timer(self.emit_period, self.auto_emit).start()
  def process(self, tup):
      # do someting
  def auto_emit(self):
      self.emit(self.xxx)
      Timer(self.emit_period, self.auto_emit).start()
```
it raised JSONDecodeError, but it works well, if i emit the tuple in the `procesee` function 
```
38793 [Thread-34] ERROR b.s.d.executor - 
java.lang.Exception: Shell Process Exception: Python JSONDecodeError raised
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/streamparse/storm/component.py", line 489, in run
    self._run()
  File "/usr/lib/python2.7/site-packages/streamparse/storm/bolt.py", line 209, in _run
    self._current_tups = [self.read_tuple()]
  File "/usr/lib/python2.7/site-packages/streamparse/storm/component.py", line 341, in read_tuple
    cmd = self.read_command()
  File "/usr/lib/python2.7/site-packages/streamparse/storm/component.py", line 334, in read_command
    msg = self.read_message()
  File "/usr/lib/python2.7/site-packages/streamparse/storm/component.py", line 315, in read_message
    return json.loads(msg)
  File "/usr/lib64/python2.7/site-packages/simplejson/__init__.py", line 516, in loads
    return _default_decoder.decode(s)
  File "/usr/lib64/python2.7/site-packages/simplejson/decoder.py", line 373, in decode
    raise JSONDecodeError("Extra data", s, end, len(s))
JSONDecodeError: Extra data: line 2 column 1 - line 3 column 1 (char 84 - 87)

	at backtype.storm.task.ShellBolt.handleError(ShellBolt.java:188) [storm-core-0.10.0.jar:0.10.0]
	at backtype.storm.task.ShellBolt.access$1100(ShellBolt.java:69) [storm-core-0.10.0.jar:0.10.0]
	at backtype.storm.task.ShellBolt$BoltReaderRunnable.run(ShellBolt.java:336) [storm-core-0.10.0.jar:0.10.0]
	at java.lang.Thread.run(Thread.java:745) [?:1.7.0_95]
38889 [Thread-34] ERROR b.s.t.ShellBolt - Halting process: ShellBolt died. Command: [python, -m, streamparse.run, bolts.tablemnt.TableMnt], ProcessInfo pid:235574, name:table-mnt-bolt exitCode:1, errorString: 
java.lang.RuntimeException: backtype.storm.multilang.NoOutputException: Pipe to subprocess seems to be broken! No output read.
Serializer Exception:
Exception in thread Thread-3:
Traceback (most recent call last):
  File "/usr/lib64/python2.7/threading.py", line 811, in __bootstrap_inner
    self.run()
  File "/usr/lib64/python2.7/threading.py", line 1083, in run
    self.function(*self.args, **self.kwargs)
  File "bolts/tablemnt.py", line 78, in auto_emit
    self.emit(('1', '2', '3'))
  File "/usr/lib/python2.7/site-packages/streamparse/storm/bolt.py", line 145, in emit
    need_task_ids=need_task_ids)
  File "/usr/lib/python2.7/site-packages/streamparse/storm/component.py", line 467, in emit
    else self.read_task_ids()
  File "/usr/lib/python2.7/site-packages/streamparse/storm/component.py", line 324, in read_task_ids
    msg = self.read_message()
  File "/usr/lib/python2.7/site-packages/streamparse/storm/component.py", line 315, in read_message
    return json.loads(msg)
  File "/usr/lib64/python2.7/site-packages/simplejson/__init__.py", line 516, in loads
    return _default_decoder.decode(s)
  File "/usr/lib64/python2.7/site-packages/simplejson/decoder.py", line 370, in decode
    obj, end = self.raw_decode(s)
  File "/usr/lib64/python2.7/site-packages/simplejson/decoder.py", line 400, in raw_decode
    return self.scan_once(s, idx=_w(s, idx).end())
JSONDecodeError: Expecting value: line 1 column 1 (char 0)



	at backtype.storm.utils.ShellProcess.readShellMsg(ShellProcess.java:101) ~[storm-core-0.10.0.jar:0.10.0]
	at backtype.storm.task.ShellBolt$BoltReaderRunnable.run(ShellBolt.java:321) [storm-core-0.10.0.jar:0.10.0]
	at java.lang.Thread.run(Thread.java:745) [?:1.7.0_95]
38889 [Thread-34] ERROR b.s.d.executor - 
java.lang.RuntimeException: backtype.storm.multilang.NoOutputException: Pipe to subprocess seems to be broken! No output read.
Serializer Exception:
Exception in thread Thread-3:
Traceback (most recent call last):
  File "/usr/lib64/python2.7/threading.py", line 811, in __bootstrap_inner
    self.run()
  File "/usr/lib64/python2.7/threading.py", line 1083, in run
    self.function(*self.args, **self.kwargs)
  File "bolts/tablemnt.py", line 78, in auto_emit
    self.emit(('1', '2', '3'))
  File "/usr/lib/python2.7/site-packages/streamparse/storm/bolt.py", line 145, in emit
    need_task_ids=need_task_ids)
  File "/usr/lib/python2.7/site-packages/streamparse/storm/component.py", line 467, in emit
    else self.read_task_ids()
  File "/usr/lib/python2.7/site-packages/streamparse/storm/component.py", line 324, in read_task_ids
    msg = self.read_message()
  File "/usr/lib/python2.7/site-packages/streamparse/storm/component.py", line 315, in read_message
    return json.loads(msg)
  File "/usr/lib64/python2.7/site-packages/simplejson/__init__.py", line 516, in loads
    return _default_decoder.decode(s)
  File "/usr/lib64/python2.7/site-packages/simplejson/decoder.py", line 370, in decode
    obj, end = self.raw_decode(s)
  File "/usr/lib64/python2.7/site-packages/simplejson/decoder.py", line 400, in raw_decode
    return self.scan_once(s, idx=_w(s, idx).end())
JSONDecodeError: Expecting value: line 1 column 1 (char 0)



	at backtype.storm.utils.ShellProcess.readShellMsg(ShellProcess.java:101) ~[storm-core-0.10.0.jar:0.10.0]
	at backtype.storm.task.ShellBolt$BoltReaderRunnable.run(ShellBolt.java:321) [storm-core-0.10.0.jar:0.10.0]
	at java.lang.Thread.run(Thread.java:745) [?:1.7.0_95]
Traceback (most recent call last):
  File "/usr/bin/sparse", line 9, in <module>
    load_entry_point('streamparse==2.1.4', 'console_scripts', 'sparse')()
  File "/usr/lib/python2.7/site-packages/streamparse/cli/sparse.py", line 57, in main
    args.func(args)
  File "/usr/lib/python2.7/site-packages/streamparse/cli/run.py", line 81, in main
    debug=args.debug)
  File "/usr/lib/python2.7/site-packages/streamparse/cli/run.py", line 52, in run_local_topology
    run(full_cmd)
  File "/usr/lib/python2.7/site-packages/invoke/__init__.py", line 27, in run
    return Context().run(command, **kwargs)
  File "/usr/lib/python2.7/site-packages/invoke/context.py", line 53, in run
    return runner_class(context=self).run(command, **kwargs)
  File "/usr/lib/python2.7/site-packages/invoke/runners.py", line 302, in run
    raise Failure(result)
invoke.exceptions.Failure: Command execution failure!

Exit code: 11

Stderr:
```

**Comment by compassur** [comment]
I find out that in the log some error message which can't be decoded: A empty string and a string mixed with command message and task ids.
it seems that the `read_message` function isn't thread safe, as you just has a lock on the execution of `input_stream.readline()` but to get a message we need to execute at least 3 `input_stream.readline()`
then, having found out this, i added a lock before `tup = self.read_tuple()` in method `_run()` and the same lock before `emit()` as each emit will end with `read_task_ids()` which solved the problem.
but i think the lock should be added on  `read_message()`, or i should pass `need_task_ids=False` to `emit()`? and what will happen if it do this, or is there any other problem?

**Comment by andrewkcarter** [comment]
I just encountered the same issue. I get the same JSONDecodeError when I started emitting and acking in a secondary thread. I tried @compassur's hack and this resolved the problem for me too.
```
    def read_message(self):
        with self._reader_lock:
            return super(Processor, self).read_message()
```

**Comment by dan-blanchard** [comment]
Before everyone piles on here, this is something that's fixed in the 3.0 developer preview releases, which is why I didn't comment earlier.  Please try upgrading to 3.0.0.dev2.  The 3.0 release is rapidly approaching, as I finally had a change to fix one of our major blockers (it was missing `sparse run` before).

**Comment by dan-blanchard** [comment]
I'm going to close this since it's fixed in 3.0, unless someone is able to replicate it there.
